src/main/java/picard/sam/markduplicates/MarkDuplicates.java

```diff
@@ -935,5 +955,42 @@ public int compare(final ReadEndsForMarkDuplicates lhs, final ReadEndsForMarkDup
             return compareDifference;
         }
     }
+    private Set<String> getNonProperlyMappedPairDuplicatesNames(long nextDuplicateIndex) {
+        final Set<String> duplicatesNames = new HashSet<>();
+        final SamHeaderAndIterator headerAndIteratorForRemoveDuplicates = openInputs(false);
+        final SAMFileHeader header = headerAndIteratorForRemoveDuplicates.header;
+        final SAMFileHeader.SortOrder sortOrder = header.getSortOrder();
+
+        long recordInFileIndex = 0;
+        String duplicateQueryName = null;
 
+        final CloseableIterator<SAMRecord> iteratorForRemoveDuplicates = headerAndIteratorForRemoveDuplicates.iterator;
+        while (iteratorForRemoveDuplicates.hasNext()) {
+            final SAMRecord rec = iteratorForRemoveDuplicates.next();
+
+            final boolean nextDuplicateIndexIsNecessary = recordInFileIndex > nextDuplicateIndex &&
+                    (sortOrder == SAMFileHeader.SortOrder.coordinate || !rec.getReadName().equals(duplicateQueryName));
+
+            if (nextDuplicateIndexIsNecessary) {
+                nextDuplicateIndex = (duplicateIndexes.hasNext() ? duplicateIndexes.next() : NO_SUCH_INDEX);
```

Comment:
Input variables shouldn't be mutated. A new variable should be introduced to hold this value.